aten/src/ATen/native/TensorShape.cpp

```diff
@@ -2777,5 +2777,264 @@ at::Tensor diagonal_scatter(const at::Tensor& self, const at::Tensor& src, int64
+
+at::Tensor& view_as_real_copy_out(const at::Tensor & self, at::Tensor & out) {
+  auto tmp = at::view_as_real(self);
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::Tensor& view_as_complex_copy_out(const at::Tensor & self, at::Tensor & out) {
+  auto tmp = at::view_as_complex(self);
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::Tensor& _conj_copy_out(const at::Tensor & self, at::Tensor & out) {
+  auto tmp = self._conj();
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::Tensor& _neg_view_copy_out(const at::Tensor & self, at::Tensor & out) {
+  auto tmp = self._neg_view();
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::Tensor& as_strided_copy_out(const at::Tensor & self, at::IntArrayRef size, at::IntArrayRef stride, c10::optional<int64_t> storage_offset, at::Tensor & out) {
+  auto tmp = self.as_strided(size, stride, storage_offset);
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::Tensor& _sparse_broadcast_to_copy_out(const at::Tensor & self, at::IntArrayRef size, at::Tensor & out) {
+  auto tmp = at::_sparse_broadcast_to(self, size);
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::Tensor& diagonal_copy_out(const at::Tensor & self, int64_t offset, int64_t dim1, int64_t dim2, at::Tensor & out) {
+  auto tmp = self.diagonal(offset, dim1, dim2);
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::Tensor& expand_copy_SymInt_out(const at::Tensor & self, c10::SymIntArrayRef size, bool implicit, at::Tensor & out) {
+  auto tmp = self.expand(size, implicit);
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::Tensor& expand_copy_out(const at::Tensor & self, at::IntArrayRef size, bool implicit, at::Tensor & out) {
+  auto tmp = self.expand(size, implicit);
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::Tensor& narrow_copy_out(const at::Tensor & self, int64_t dim, int64_t start, int64_t length, at::Tensor & out) {
+  auto tmp = self.narrow(dim, start, length);
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::Tensor& permute_copy_out(const at::Tensor & self, at::IntArrayRef dims, at::Tensor & out) {
+  auto tmp = self.permute(dims);
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::Tensor& _reshape_alias_copy_out(const at::Tensor & self, at::IntArrayRef size, at::IntArrayRef stride, at::Tensor & out) {
+  auto tmp = self._reshape_alias(size, stride);
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::Tensor& select_copy_int_out(const at::Tensor & self, int64_t dim, int64_t index, at::Tensor & out) {
+  auto tmp = self.select(dim, index);
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::Tensor& detach_copy_out(const at::Tensor & self, at::Tensor & out) {
+  auto tmp = self.detach();
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::Tensor& slice_copy_Tensor_out(const at::Tensor & self, int64_t dim, c10::optional<int64_t> start, c10::optional<int64_t> end, int64_t step, at::Tensor & out) {
+  auto tmp = self.slice(dim, start, end, step);
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::ITensorListRef split_copy_Tensor_out(const at::Tensor & self, int64_t split_size, int64_t dim, at::TensorList  out) {
```

Comment:
@ezyang / @ysiraichi This PR isn't ready for review yet, but I'm wondering - what do you think the dispatcher return type should be for functions like this, with the following schema?
```
split_copy.Tensor_out(Tensor self, int split_size, int dim=0, *, Tensor(a!)[] out) -> Tensor(a!)[]
```

tldr is that this PR introduces some new ops that have a schema that the codegen hasn't had to deal with before - out= arguments that are lists of mutable tensors. I think the answer is that we want it to be `ITensorListRef`, but that currently breaks - so I wanted to clarify my reasoning.

Just to list out all the options:

(1) make the return type `ITensorListRef`

This breaks boxed_to_unboxed code right now, because we don't currently know how to convert `IValue` to `ITensorListRef`. I think... this is probably what we want? But it looks like this doesn't work today, so I was wondering what the current state of `ITensorListRef` is (or maybe I'm misunderstanding something - shouldn't you be able to convert an `IValue` to an `ITensorListRef`?)

(2) make the return type `std::vector<Tensor>`

This will work, but has two disadvantages: We have a bunch of codegen that expects the out= argument and the return type to be the same, so that we can just `return out` in the generated kernels. We'd have to modify those to explicitly convert the output to a vector (`return out.vec()`), and also accept the perf hit (I don't expect any of these kernels to be hotpath though).

(3) make the return type `at::TensorList`

This also breaks today, because `IValue`s don't know how to convert to a `TensorList` (`ArrayRef<Tensor>`). I think it would also be... wrong, because we'd need to convert from the IValue's internal `c10::List<Tensor>` to `ArrayRef<Tensor>`, which requires creating new tensors to make them all contiguous in memory, but they aren't owned by anyone.

Reply:
I was just discussing this at https://github.com/pytorch/pytorch/pull/76049 We should make the return type void. This would be BC breaking but I doubt anyone is actually using this API and in any case they can't do method chaining like this.